File: go/levenshtein.go
package main

import (
  "os"
  "fmt"
  // "math"
)

// penalties
const (
  EQ = 1
  INS = 10
  SUB = 5
  DEL = 10
)

type XY struct {
  X int
  Y int
}

func Minimum(a, b int) int {
  if (a < b) {
    return a
  }
  return b
}

func MinEditDistance(source string, target string) int {
  memo := make([][]int, len(source) + 1)
  for s := range memo {
    memo[s] = make([]int, len(target) + 1)
    for t := range memo[s] {
      memo[s][t] = -1
    }
    // initialize the left border, where the origin is at the bottom left.
    memo[s][0] = s
  }
  // initialize the bottom border, where the origin is at the bottom left.
  for t, _ := range target {
    memo[0][t + 1] = t + 1
  }

  var dist func(s int, t int) int
  dist = func(s int, t int) int {
    min := memo[s][t]
    if min == -1 {
      // fmt.Printf("%s[%d] - %s[%d]\n", source, s, target, t)
      if source[s - 1] == target[t - 1] {
        min = dist(s - 1, t - 1) + EQ
      } else {
        // deletion is like removing one from the source
        del := dist(s - 1, t) + DEL
        // insertion is like adding one to the target
        ins := dist(s, t - 1) + INS
        sub := dist(s - 1, t - 1) + SUB
        min = Minimum(del, Minimum(ins, sub))
      }
      memo[s][t] = min
    }
    return min
  }
  return dist(len(source), len(target))
}


func main() {
  source := os.Args[1]
  target := os.Args[2]
  fmt.Println(source + " -> " + target)
  fmt.Printf("MinEditDistance: %d\n", MinEditDistance(source, target))
}

File: javascript/levenshtein.js
var levenshtein_matrix = function(source, target) {
  var nrows = source.length + 1;
  // var source_chars = source.split('');
  var ncols = target.length + 1;
  // var target_chars = target.split('');
  // initialize to something like this:
  //     c a t
  //   0 1 2 3
  // m 1 0 0 0
  // a 2 0 0 0
  // t 3 0 0 0
  //
  // i.e.,
  //   distances = [[0, 1, 2, 3], [1, 0, 0, 0], [2, 0, 0, 0], [3, 0, 0, 0]];
  //   (the primary index is by row = by source character)
  var i; // i is for rows
  var j; // j is for cols
  var distances = [];
  for (i = 0; i < nrows; i++) {
    distances.push([]);
    for (j = 0; j < ncols; j++) {
      var cell = 0;
      // top row
      if (i === 0) cell = j;
      // left column
      if (j === 0) cell = i;
      distances[i][j] = cell;
    }
  }
  // calculate distances
  for (i = 1; i < nrows; i++) {
    for (j = 1; j < ncols; j++) {
      if (source[i-1] === target[j-1]) {
        distances[i][j] = distances[i-1][j-1]; // equal (no cost)
      }
      else {
        distances[i][j] = Math.min(
          distances[i-1][j] + 1, // a deletion
          distances[i][j-1] + 1, // an insertion
          distances[i-1][j-1] + 1 // a substitution
        );
      }
    }
  }
  return distances;
};
